sys/net/application_layer/nanocoap/nanocoap.c

```diff
@@ -439,6 +439,41 @@ int coap_get_blockopt(coap_pkt_t *pkt, uint16_t option, uint32_t *blknum, uint8_
     return (blkopt & 0x8) ? 1 : 0;
 }
 
+bool coap_find_uri_query(coap_pkt_t *pkt, const char *key, const char **value, size_t *len)
+{
+    uint8_t *opt_pos = NULL;
+
+    while (1) {
+        int len_query;
+        const void *key_data = coap_iterate_option(pkt, COAP_OPT_URI_QUERY,
+                                                   &opt_pos, &len_query);
+        if (key_data == NULL) {
+            return false;
+        }
+
+        const char *value_data = memchr(key_data, '=', len_query);
+
+        if (memcmp(key, key_data, value_data ? value_data - (char *)key_data : len_query)) {
+            continue;
+        }
+
+        if (value) {
+            *value = value_data
+                   ? value_data + 1
+                   : NULL;
+        }
+        if (len) {
+            size_t len_key = value_data
+                           ? *value - (char *)key_data
+                           : len_query;
+            *len = len_query - len_key;
+        }
```

Comment:
This will crash if `value` is `null`, but `len` (and `value_data`) isn't. Maybe also rename `value_data` to something more meaningful such as `key_value_separator`.

I would suggest calculating `len_key` directly below `memchr` and reuse it for both `memcmp` and the value `len` calculation here.

Reply:
True, let's get rid of that condition entirely - if you get the value you must also get the length as it is not zero-terminated. And the length without the value is of little use.